File: python/a28.py
# A28 Blackboard
# submission: https://atcoder.jp/contests/tessoku-book/submissions/38567348

MODULO = 1_00_00

class Modulo:
    def __init__(self, value: int, modulo: int):
        self.value = value
        self.modulo = modulo
    def __str__(self) -> str:
        return str(self.value)
    def __iadd__(self, other):
        if not type(other) is Modulo or self.modulo != other.modulo:
            raise TypeError()
        self.value = (self.value + other.value) % self.modulo
        return self
    def __isub__(self, other):
        if not type(other) is Modulo or self.modulo != other.modulo:
            raise TypeError()
        self.value = (self.value - other.value) % self.modulo
        return self
    def __imul__(self, other):
        if not type(other) is Modulo or self.modulo != other.modulo:
            raise TypeError()
        self.value = self.value * other.value % self.modulo
        return self

n = int(input())

b = Modulo(0, MODULO)

for i in range(n):
    t, a = input().split()
    a = Modulo(int(a), MODULO)
    if t == '+':
        b += a
    if t == '-':
        b -= a
    if t == '*':
        b *= a
    print(b)

File: rust/tessoku-book/src/bin/a28.rs
// A28 Blackboard
// submission: https://atcoder.jp/contests/tessoku-book/submissions/38568646

use proconio::input;
use std::ops::{AddAssign, SubAssign, MulAssign};

const MODULO: u32 = 1_00_00;

struct Modulo {
    value: u32,
    modulo: u32,
}

impl AddAssign for Modulo {
    fn add_assign(&mut self, other: Self) {
        if self.modulo != other.modulo {
            panic!("modulo isn't the same");
        }
        *self = Self {
            value: (self.value + other.value) % self.modulo,
            modulo: self.modulo,
        };
    }
}

impl SubAssign for Modulo {
    fn sub_assign(&mut self, other: Self) {
        if self.modulo != other.modulo {
            panic!("modulo isn't the same");
        }
        *self = Self {
            value: (self.modulo + self.value - other.value) % self.modulo,
            modulo: self.modulo,
        };
    }
}

impl MulAssign for Modulo {
    fn mul_assign(&mut self, other: Self) {
        if self.modulo != other.modulo {
            panic!("modulo isn't the same");
        }
        *self = Self {
            value: (self.value * other.value) % self.modulo,
            modulo: self.modulo,
        };
    }
}

fn main() {
    input!{
        n: usize,
    }
    let mut b = Modulo{value: 0, modulo: MODULO};
    for _n in 0..n {
        input!{
            t: char,
            a: u32,
        }
        let aa = Modulo{value: a, modulo: MODULO};
        match t {
            '+' => b += aa,
            '-' => b -= aa,
            '*' => b *= aa,
            _ => panic!(),
        }
        println!("{}", b.value);
    }
}
